compiler/rustc_codegen_llvm/src/common.rs

```diff
@@ -255,21 +255,38 @@ impl<'ll, 'tcx> ConstMethods<'tcx> for CodegenCx<'ll, 'tcx> {
                 let (prov, offset) = ptr.into_parts();
                 let (base_addr, base_addr_space) = match self.tcx.global_alloc(prov.alloc_id()) {
                     GlobalAlloc::Memory(alloc) => {
-                        let init = const_alloc_to_llvm(self, alloc);
-                        let alloc = alloc.inner();
-                        let value = match alloc.mutability {
-                            Mutability::Mut => self.static_addr_of_mut(init, alloc.align, None),
-                            _ => self.static_addr_of(init, alloc.align, None),
-                        };
-                        if !self.sess().fewer_names() && llvm::get_value_name(value).is_empty() {
-                            let hash = self.tcx.with_stable_hashing_context(|mut hcx| {
-                                let mut hasher = StableHasher::new();
-                                alloc.hash_stable(&mut hcx, &mut hasher);
-                                hasher.finish::<Hash128>()
-                            });
-                            llvm::set_value_name(value, format!("alloc_{hash:032x}").as_bytes());
+                        // For ZSTs directly codegen an aligned pointer.
+                        // This avoids generating a zero-sized constant value and actually needing a
+                        // real address at runtime.
+                        if alloc.inner().len() == 0 {
+                            assert_eq!(offset.bytes(), 0);
+                            let llval = self.const_usize(alloc.inner().align.bytes());
```

Comment:
This is dangerous. If there is any other point to the same allocation anywhere else that does not go through this codepath, then we would now be generating two different addresses for the same pointer.

So if you do this then please also add an assertion in `const_alloc_to_llvm` that it is never called on zero-sized allocations, to make sure we do not accidentally have both a "real" and a "fake" address for these allocations.

Reply:
`const_alloc_to_llvm` is also used by `codegen_static_initializer`.

So if there's ever a `Scalar::Ptr` to a `GlobalAlloc::Memory` that is actually a static, then we'd use inconsistent addresses for that static.

Reply:
> So if there's ever a Scalar::Ptr to a GlobalAlloc::Memory that is actually a static, then we'd use inconsistent addresses for that static.

I think the case you're describing is already broken? The code path prior to this PR would already generate two distinct allocations -- at least, I can't see anything that deduplicates between statics and consts. We have a `const_globals` map that maps an LLVM const value to an LLVM global, but statics as far as I can tell don't use that map, they only get deduplicated based on DefId.

I'm happy to change `const_alloc_to_llvm` to apply this optimization there but I'm not convinced that it's a complete fix in any case if we really need this for correctness (e.g., a cranelift + LLVM library would be wrong, presumably? Not sure that's supported). I guess that implies that the MIR-level optimization may be a better idea here as you point out below...

FWIW, it looks like cranelift is doing something like this optimization already: https://github.com/rust-lang/rust/blob/ccfcd950b333fed046275dd8d54fe736ca498aa7/compiler/rustc_codegen_cranelift/src/constant.rs#L103. It looks to me like that is also not shared between statics and constants.

I have not looked at gcc yet.

Reply:
> I think the case you're describing is already broken? The code path prior to this PR would already generate two distinct allocations -- at least, I can't see anything that deduplicates between statics and consts. We have a const_globals map that maps an LLVM const value to an LLVM global, but statics as far as I can tell don't use that map, they only get deduplicated based on DefId.

Hm, that is true. And since @oli-obk's recent changes, there generally cannot exist a `GlobalAlloc::Memory` for the data backing a static. So we should be good here.

But still, `const_alloc_to_llvm` should get some sort of safeguard to make sure zero-sized allocs are treated consistently. In its current form it can't just assert as that would fail for zero-sized statics; maybe the code that calls `const_alloc_to_llvm` for static should have a zero-sized case as well and then `const_alloc_to_llvm` can get an assertion?

> FWIW, it looks like cranelift is doing something like this optimization already: 

I think that's different; that's for generating *values* that are zero-sized whereas you are changing the code for *pointers to zero-sized const-allocations*. Values don't have an address so there it's trivially correct to do what the cranelift backend does.

Reply:
> But still, const_alloc_to_llvm should get some sort of safeguard to make sure zero-sized allocs are treated consistently. In its current form it can't just assert as that would fail for zero-sized statics; maybe the code that calls const_alloc_to_llvm for static should have a zero-sized case as well and then const_alloc_to_llvm can get an assertion?

It looks to me like doing this for statics would be much harder. Those are treated considerably differently in the code to the ephemeral consts I'm currently modifying -- we'd need to avoid *declaring* the ZST static entirely I think, which happens separately from providing an initializer for it. And that would need to be only for locally defined statics (e.g., `extern { static BAR: (); }` probably cannot be erased from existence...; even for local statics something like `#[used]` has unclear implications if we're deleting ZST statics.

I think it would be possible to avoid changing the static declaration code at all (and adding any asserts to const_alloc_to_llvm) but still uniformly generate a constant rather than a global reference in the *pointer to static* lowering (i.e., same function currently being modified). LLVM would probably be able to drop the useless statics we generated elsewhere in that case.

Reply:
Hm yeah, makes sense.

Maybe the best we can do is just tell `const_alloc_to_llvm` whether it is translating a static. I don't think it has any way to determine that itself?

Reply:
What would we do inside it though? It's generating the value of the static, not the pointer?

Do you mean so that we can assert we only generate ZST allocations for statics?

Reply:
> Do you mean so that we can assert we only generate ZST allocations for statics?

Yes, that.

Reply:
OK, should be implemented.